fix: enums are numbers, not strings

eqyiel opened the pull request:
[Protobuf says](https://developers.google.com/protocol-buffers/docs/proto3#enum)

> Enumerator constants must be in the range of a 32-bit integer. 

It's the same for [proto2](https://developers.google.com/protocol-buffers/docs/proto).

With the [default configuration](https://github.com/sue71/protoc-gen-jsonpb-ts/blob/80800daa2c2ed3f49cf344dc7ba7711e1697a004/src/config.ts#L52), this [sample file](https://github.com/googleapis/googleapis/blob/dc65f39f33cb139b3a2244199a3e722a4d94b679/google/rpc/code.proto#L25-L186) is compiled incorrectly.

The [README](https://github.com/sue71/protoc-gen-jsonpb-ts/blob/80800daa2c2ed3f49cf344dc7ba7711e1697a004/README.md) says

> Generate TypeScript definitions according to the spec. 

Therefore, we should fix the behaviour.

This is a breaking change for clients that depend on the non-spec compliant behaviour.

### Current behaviour

On 80800da:

```
~/git/googleapis master
❯ git log --oneline -1
dc65f39f3 (HEAD -> master, origin/master, origin/HEAD) For Secret Manager v1 and v1beta1, noted Secret ID character limitations.

~/git/mercari/googleapis master*
❯ (cd ../protoc-gen-jsonpb-ts; git log --oneline -1)
80800da (HEAD -> master, origin/master, origin/HEAD) Merge pull request #36 from sue71/renovate/prettier-2.x

~/git/googleapis master
❯ protoc --version
libprotoc 3.7.1

~/git/googleapis master*
❯ node --version
v10.16.0

~/git/googleapis master
❯ test -d ./out && rm -rf ./out; mkdir ./out; protoc -I . --plugin=../protoc-gen-jsonpb-ts/bin/protoc-gen-jsonpb-ts --jsonpb-ts_out="ignorePackage=true:./out" google/rpc/code.proto; cat ./out/google/rpc/code_pb.ts
/* GENERATED FROM google/rpc/code.proto. DO NOT EDIT MANUALLY. */
/* tslint:disabled */
/* eslint-disable */

export enum Code {
  OK = "OK",
  CANCELLED = "CANCELLED",
  UNKNOWN = "UNKNOWN",
  INVALID_ARGUMENT = "INVALID_ARGUMENT",
  DEADLINE_EXCEEDED = "DEADLINE_EXCEEDED",
  NOT_FOUND = "NOT_FOUND",
  ALREADY_EXISTS = "ALREADY_EXISTS",
  PERMISSION_DENIED = "PERMISSION_DENIED",
  UNAUTHENTICATED = "UNAUTHENTICATED",
  RESOURCE_EXHAUSTED = "RESOURCE_EXHAUSTED",
  FAILED_PRECONDITION = "FAILED_PRECONDITION",
  ABORTED = "ABORTED",
  OUT_OF_RANGE = "OUT_OF_RANGE",
  UNIMPLEMENTED = "UNIMPLEMENTED",
  INTERNAL = "INTERNAL",
  UNAVAILABLE = "UNAVAILABLE",
  DATA_LOSS = "DATA_LOSS",
}

~/git/googleapis master*
❯ test -d ./out && rm -rf ./out; mkdir ./out; protoc -I . --plugin=../protoc-gen-jsonpb-ts/bin/protoc-gen-jsonpb-ts --jsonpb-ts_out="ignorePackage=true,enumFormat=stringLiteral:./out" google/rpc/code.proto; cat ./out/google/rpc/code_pb.ts

/* GENERATED FROM google/rpc/code.proto. DO NOT EDIT MANUALLY. */
/* tslint:disabled */
/* eslint-disable */

export type Code =
  | "OK"
  | "CANCELLED"
  | "UNKNOWN"
  | "INVALID_ARGUMENT"
  | "DEADLINE_EXCEEDED"
  | "NOT_FOUND"
  | "ALREADY_EXISTS"
  | "PERMISSION_DENIED"
  | "UNAUTHENTICATED"
  | "RESOURCE_EXHAUSTED"
  | "FAILED_PRECONDITION"
  | "ABORTED"
  | "OUT_OF_RANGE"
  | "UNIMPLEMENTED"
  | "INTERNAL"
  | "UNAVAILABLE"
  | "DATA_LOSS";

```

### Expected behaviour (fixed in this PR)

```
~/git//googleapis master
❯ git log --oneline -1
dc65f39f3 (HEAD -> master, origin/master, origin/HEAD) For Secret Manager v1 and v1beta1, noted Secret ID character limitations.

~/git//googleapis master*
❯ (cd ../protoc-gen-jsonpb-ts; git log --oneline -1)
ca3d056 (HEAD -> fix/enums-are-numbers, eqyiel/fix/enums-are-numbers) fix: also use number type when "stringLiteral" option is passed

~/git/googleapis master
❯ protoc --version
libprotoc 3.7.1

~/git/googleapis master*
❯ node --version
v10.16.0

~/git/googleapis master
❯ test -d ./out && rm -rf ./out; mkdir ./out; protoc -I . --plugin=../protoc-gen-jsonpb-ts/bin/protoc-gen-jsonpb-ts --jsonpb-ts_out="ignorePackage=true:./out" google/rpc/code.proto; cat ./out/google/rpc/code_pb.ts
/* GENERATED FROM google/rpc/code.proto. DO NOT EDIT MANUALLY. */
/* tslint:disabled */
/* eslint-disable */

export enum Code {
  OK = 0,
  CANCELLED = 1,
  UNKNOWN = 2,
  INVALID_ARGUMENT = 3,
  DEADLINE_EXCEEDED = 4,
  NOT_FOUND = 5,
  ALREADY_EXISTS = 6,
  PERMISSION_DENIED = 7,
  UNAUTHENTICATED = 16,
  RESOURCE_EXHAUSTED = 8,
  FAILED_PRECONDITION = 9,
  ABORTED = 10,
  OUT_OF_RANGE = 11,
  UNIMPLEMENTED = 12,
  INTERNAL = 13,
  UNAVAILABLE = 14,
  DATA_LOSS = 15,
}

~/git/googleapis master*
❯ test -d ./out && rm -rf ./out; mkdir ./out; protoc -I . --plugin=../protoc-gen-jsonpb-ts/bin/protoc-gen-jsonpb-ts --jsonpb-ts_out="ignorePackage=true,enumFormat=stringLiteral:./out" google/rpc/code.proto; cat ./out/google/rpc/code_pb.ts

/* GENERATED FROM google/rpc/code.proto. DO NOT EDIT MANUALLY. */
/* tslint:disabled */
/* eslint-disable */

export type Code =
  | 0
  | 1
  | 2
  | 3
  | 4
  | 5
  | 6
  | 7
  | 16
  | 8
  | 9
  | 10
  | 11
  | 12
  | 13
  | 14
  | 15;
```

eqyiel committed:
fix: enums are numbers, not strings

eqyiel committed:
fix: also use number type when "stringLiteral" option is passed

eqyiel commented:
I think the `stringLiteral` option should be renamed because it doesn't make sense (it didn't make sense before either.

Before: it outputs  a _string union_ type.
Now: it outputs a _number union_ type.

https://www.typescriptlang.org/docs/handbook/advanced-types.html#union-types

sue71 commented:
@eqyiel  Thanks for PR 👍 

This library generates code according to the specification of  "jsonMapping". According to the specification, enum is mapped to string, so the current output is correct.
see https://developers.google.com/protocol-buffers/docs/proto3#json

> I think the stringLiteral option should be renamed because it doesn't make sense (it didn't make sense before either.

Your naming is nice, but I think enumFormat.stringLiteral fully describes the specification.
I don't think we need to make breaking changes to rename it.

https://www.typescriptlang.org/docs/handbook/advanced-types.html#string-literal-types

eqyiel commented:
@sue71 I see, thank you for telling me, I didn't know the default specification for JSON Mapping!  I was so sure that this behaviour was a mistake because I can't find a use case for it.  I was surprised to learn that similar libraries have the proposed behaviour by default.

<details>

<summary>improbable-eng/ts-protoc-gen example</summary>

```
~/git/googleapis master*
❯ test -d ./out && rm -rf ./out; mkdir ./out; protoc -I . --plugin=../protoc-gen-ts --ts_out=./out google/rpc/code.proto; cat ./out/google/rpc/code_pb.d.ts
removed './out/google/rpc/code_pb.d.ts'
removed directory './out/google/rpc'
removed directory './out/google'
removed directory './out'
// package: google.rpc
// file: google/rpc/code.proto

import * as jspb from "google-protobuf";

export interface CodeMap {
  OK: 0;
  CANCELLED: 1;
  UNKNOWN: 2;
  INVALID_ARGUMENT: 3;
  DEADLINE_EXCEEDED: 4;
  NOT_FOUND: 5;
  ALREADY_EXISTS: 6;
  PERMISSION_DENIED: 7;
  UNAUTHENTICATED: 16;
  RESOURCE_EXHAUSTED: 8;
  FAILED_PRECONDITION: 9;
  ABORTED: 10;
  OUT_OF_RANGE: 11;
  UNIMPLEMENTED: 12;
  INTERNAL: 13;
  UNAVAILABLE: 14;
  DATA_LOSS: 15;
}

export const Code: CodeMap;
```

</details>


`enumFormat.stringLiteral` makes sense to me now.

However

- https://godoc.org/github.com/golang/protobuf/jsonpb#Marshaler has an `EnumsAsInts` option
- https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.util.json_util has an `always_print_enums_as_ints` option
- https://github.com/protocolbuffers/protobuf/blob/3be9322e28dfcacdb669f646f76fe833fa274107/csharp/src/Google.Protobuf/JsonFormatter.cs#L806-L809 has `FormatEnumsAsIntegers`
- https://googleapis.dev/python/protobuf/latest/google/protobuf/json_format.html#google.protobuf.json_format.MessageToDict has `use_integers_for_enums`

I guess what I want to add is something like this.  So I have a new proposal.

@sue71 Would you accept a PR that adds

- a new enum format `enumFormat.numericLiteral` (like `numeric literal` in the TS docs https://www.typescriptlang.org/docs/handbook/advanced-types.html#numeric-literal-types)

  The output of `google/rpc/code.proto` would be like this:

  ```
  export type Code =
    | 0
    | 1
    | 2
    | 3
    | 4
    | 5
    | 6
    | 7
    | 16
    | 8
    | 9
    | 10
    | 11
    | 12
    | 13
    | 14
    | 15;
  ```
  
- a new enum format `enumFormat.numericEnum` (I don't know if the name is good)

  The output of `google/rpc/code.proto` would be like this:

  ```
  export enum Code {
    OK = 0,
    CANCELLED = 1,
    UNKNOWN = 2,
    INVALID_ARGUMENT = 3,
    DEADLINE_EXCEEDED = 4,
    NOT_FOUND = 5,
    ALREADY_EXISTS = 6,
    PERMISSION_DENIED = 7,
    UNAUTHENTICATED = 16,
    RESOURCE_EXHAUSTED = 8,
    FAILED_PRECONDITION = 9,
    ABORTED = 10,
    OUT_OF_RANGE = 11,
    UNIMPLEMENTED = 12,
    INTERNAL = 13,
    UNAVAILABLE = 14,
    DATA_LOSS = 15,
  }
  ```
  
  - `enumFormat.enum` and `enumFormat.stringLiteral` are not changed

  - The default `enumFormat` is still `enumFormat.enum`

  - There are no breaking changes and I will be happy because I can continue to use this library ❤️

eqyiel commented:
@sue71 would you accept a PR that adds `enumFormat.numericLiteral` and `enumFormat.numericEnum` like that? ☝️

sue71 commented:
@eqyiel I'm sorry it's so late.
Thank you for the great suggestions. I also checked the specifications for other libraries. I'm waiting for the PR.

eqyiel commented:
@sue71 I decided not to work on this, sorry.

There are two reasons.

- There is https://github.com/sue71/protoc-gen-jsonpb-ts/issues/43, it seems difficult to fix, and I'm not sure if it _can_ be fixed without breaking the features which make this library useful for you :cry:
- There is https://github.com/stephenh/ts-proto, which does almost the same thing, is more actively maintained right now, and already supports "enums as numbers".  If you use compiler flags `--ts_proto_out="useOptionals=true,outputEncodeMethods=false,outputJsonMethods=false,outputClientImpl=false,forceLong=string:."` you can get output very similar to this library ("just types" with no client implementation).

Thanks for your support and teaching me about the specification of `jsonMapping`. 🙇‍♂️
